File: DFS_BFS/466-DFS-CombinationSum3.py
"""
Find all valid combinations of k numbers that sum up to n such that the following conditions are true:
Only numbers 1 through 9 are used.
Each number is used at most once.
Return a list of all possible valid combinations. The list must not contain the same combination twice, 
and the combinations may be returned in any order.
"""
from itertools import combinations,permutations
from collections import deque
def simpleMethod(k,n):
    return [x for x in list(combinations(range(1,10),k)) if sum(x)==n]
def combinationSum3(k, n):
    """
    :type k: int
    :type n: int
    :rtype: List[List[int]]
    """
    # should determine the range of k and n is reasonable first
    if k>9 or n>45:
        return []
    ans=[]
    def f(cur,i,s):
        if len(cur)>k or s>n:
            return 
        if len(cur)==k and s==n:
            ans.append(cur[:])  #here should be deep copy 
        
        for x in range(i,10):
            f(cur+[x],x+1,s+x)  #here should be x+1 for next number
    f([],1,0)
    return ans 
            
        


print(combinationSum3(3,9))
print(simpleMethod(3,9))

File: DFS_BFS/466-DFS-CombinationSum3.cpp
#include<vector>
#include<iostream>

using VEC=std::vector<int>;
using MATRIX=std::vector<VEC>;

void f(MATRIX& m,VEC cur,int x,int s,int n,int k)
{
    if((cur.size()>k) || s>n)
        return ;
    if((cur.size()==k) && (s==n))
        m.push_back(cur);
    for(int i=x;i<10;++i)
    {
        VEC t{cur.begin(),cur.end()};  //here we should pass one temporary vector 
        t.push_back(i);       
        // cur.push_back(i);
        f(m,t,i+1,s+i,n,k);
    }
}

MATRIX combinationSum(int n,int k)
{
    MATRIX m{};
    if(n>45 || k>9)
        return m;
    VEC cur{};
    f(m,cur,1,0,n,k);
    return m;
}

int main()
{
    MATRIX ans=combinationSum(9,3);
    for(auto &v:ans)
    {
        std::cout<<'\n';
        for(auto &d:v)
            std::cout<<d<<',';
    }
}